substrate/emissions/pallet/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+
+        // update the current volume
+        LastSwapVolume::<T>::set(n, Some(*vol));
+
+        total_volume = total_volume.saturating_add(vol_this_epoch);
+        *vol = vol_this_epoch;
+      }
+
+      // map epoch ec-security-distance/volume to rewards
+      let rewards_per_network = if pre_ec_security {
+        distances
+          .into_iter()
+          .map(|(n, distance)| {
+            // calculate how much each network gets based on distance to ec-security
+            let reward = u64::try_from(
+              u128::from(reward_this_epoch).saturating_mul(u128::from(distance)) /
+                u128::from(total_distance),
+            )
+            .unwrap();
+            (n, reward)
+          })
+          .collect::<BTreeMap<NetworkId, u64>>()
+      } else {
+        volume_per_network
+          .into_iter()
+          .map(|(n, vol)| {
+            // 20% of the reward goes to the Serai network and rest is distributed among others
+            // based on swap-volume.
+            let reward = if n == NetworkId::Serai {
+              reward_this_epoch / 5
+            } else {
+              let reward = reward_this_epoch - (reward_this_epoch / 5);
+              // TODO: It is highly unlikely but what to do in case of 0 total volume?
+              if total_volume != 0 {
+                u64::try_from(
+                  u128::from(reward).saturating_mul(u128::from(vol)) / u128::from(total_volume),
+                )
+                .unwrap()
+              } else {
+                0
+              }
+            };
+            (n, reward)
+          })
+          .collect::<BTreeMap<NetworkId, u64>>()
+      };
+
+      // distribute the rewards within the network
+      for (n, reward) in rewards_per_network {
+        let (validators_reward, pool_reward) = if n == NetworkId::Serai {
+          (reward, 0)
+        } else {
+          // calculate pool vs validator share
+          let capacity = ValidatorSets::<T>::total_allocated_stake(n).unwrap_or(Amount(0)).0;
+          let required = ValidatorSets::<T>::required_stake_for_network(n);
+          let unused_capacity = capacity.saturating_sub(required);
+
+          let distribution = unused_capacity.saturating_mul(ACCURACY_MULTIPLIER) / capacity;
+          let total = DESIRED_DISTRIBUTION.saturating_add(distribution);
+
+          let validators_reward = DESIRED_DISTRIBUTION.saturating_mul(reward) / total;
+          let pool_reward = reward.saturating_sub(validators_reward);
+          (validators_reward, pool_reward)
+        };
+
+        // distribute validators rewards
+        if Self::distribute_to_validators(n, validators_reward).is_err() {
+          // TODO: log the failure
+          continue;
+        }
+
+        // send the rest to the pool
+        let coin_count = u64::try_from(n.coins().len()).unwrap();
+        for c in n.coins() {
+          // assumes reward is equally distributed between network coins.
```

Comment:
We should calculate per-pool, and simply decide the validator split off the sum for the network.